File: backend-node/src/services/PaymentService.ts
import { InternalServerError } from "routing-controllers";

import PlanRepository from "../repository/PlanRepository";
import PaymentRepository from "../repository/PaymentRepository";
import BusinessRepository from "../repository/BusinessRepository";
import Business from "../entity/Business";

import { client } from "../mercadopago"
import { Preference } from "mercadopago";
import Plan from "../entity/Plan";
import Payment from "../entity/Payment";
import NotFound from "../errors/NotFound";

import { format } from "date-fns";


export default class PaymentService{

    private planRepository: PlanRepository = new PlanRepository();
    private paymentRepository: PaymentRepository = new PaymentRepository();

    private businessRepository: BusinessRepository = new BusinessRepository();
    

    public async createCheckoutPro(business: Business, origin: string, plan: number, testId: number){
        const currentPlan = await this.planRepository.findOnePlanById(plan);

        const userEmail = business.getLogin().getEmail();

        try {
            const preference = new Preference(client);

            const createdPreference = await preference.create({
                body: {
                external_reference: `${testId}`,
                metadata: {
                    id: business.getId(),
                    userEmail: userEmail,
                    plan: plan,
                },
                ...(userEmail && {
                    payer: {
                    email: userEmail,
                    },
                }),
                items: [
                    {
                    id: `${currentPlan.getId()}`,
                    description: currentPlan.getDescription(),
                    title: currentPlan.getName(),
                    quantity: 1,
                    unit_price: currentPlan.getPrice(),
                    currency_id: "BRL",
                    category_id: "category",
                    },
                ],
                payment_methods: {
                    installments: 12,
                },
                auto_return: "approved",
                back_urls: {
                    success: `${origin}/?status=sucesso`,
                    failure: `${origin}/?status=falha`,
                    pending: `${origin}/api/mercado-pago/pending`,
                },
                },
            });

            if (!createdPreference.id) {
                throw new Error("No preferenceID");
            }

            await this.savePayment(business, plan);

            return JSON.stringify({
                preferenceId: createdPreference.id,
                initPoint: createdPreference.init_point,
            });

            } catch (err) {
                console.log(err);
                throw new InternalServerError("Internal Server Error");
        }
    }

    public async updatePayment(data: any){
        console.log(data);

        const payment: any = await this.paymentRepository.findOnePaymentByBusinessId(data.personId);

        if(!payment){
            console.log("Payment not found");
            throw new NotFound("Payment not found");
        }

        const confirmedDatePayment = new Date();

        data["confirmedDatePayment"] = format(confirmedDatePayment, "yyyy-MM-dd HH:mm:ss");

        const updatedPayment = await this.paymentRepository.updatePayment(data, payment.id);

        console.log("Update: \n" + updatedPayment);

        return updatedPayment;

        //Ou aqui pode colocar a parte de enviar o email
    }

    private async savePayment(business: Business, planId: number){
        const plan: Plan = await this.planRepository.findOnePlanById(planId);

        const currentPayment: Payment = new Payment(undefined, undefined, "unpaid", undefined, undefined, undefined, undefined, undefined, plan, business);
    
        await this.paymentRepository.save(currentPayment);
    }
}

File: backend-java/src/main/java/kafka/system/br/AutoFrota/service/PaymentService.java
package kafka.system.br.AutoFrota.service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import com.mercadopago.client.payment.PaymentClient;
import com.mercadopago.client.preference.PreferenceBackUrlsRequest;
import com.mercadopago.client.preference.PreferenceClient;
import com.mercadopago.client.preference.PreferenceItemRequest;
import com.mercadopago.client.preference.PreferencePayerRequest;
import com.mercadopago.client.preference.PreferenceRequest;
import com.mercadopago.exceptions.MPApiException;
import com.mercadopago.exceptions.MPException;
import com.mercadopago.resources.preference.Preference;
import com.mercadopago.resources.preference.PreferenceBackUrls;

import kafka.system.br.AutoFrota.dto.MercadoPagoDTO;
import kafka.system.br.AutoFrota.dto.RegisterDTO;
import kafka.system.br.AutoFrota.exception.MercadoPagoException;
import kafka.system.br.AutoFrota.exception.NotFoundEntityException;
import kafka.system.br.AutoFrota.exception.PasswordIsNotConfirmedException;
import kafka.system.br.AutoFrota.model.Company;
import kafka.system.br.AutoFrota.model.Login;
import kafka.system.br.AutoFrota.model.Payment;
import kafka.system.br.AutoFrota.model.Plan;
import kafka.system.br.AutoFrota.repository.CompanyRepository;
import kafka.system.br.AutoFrota.repository.LoginRepository;
import kafka.system.br.AutoFrota.repository.PaymentRepository;
import kafka.system.br.AutoFrota.repository.PlanRepository;
import kafka.system.br.AutoFrota.utils.Email;

@Service
public class PaymentService {

    @Autowired
    private PlanRepository planRepository;

    @Autowired
    private PaymentRepository paymentRepository;

    @Autowired
    private CompanyRepository companyRepository;

    @Autowired
    private LoginRepository loginRepository;

    @Autowired
    private Email emailUtil;

    public MercadoPagoDTO createCheckoutPro(RegisterDTO dto, Company company){

        Plan currentPlan = planRepository.findById(dto.planId()).orElseThrow(() -> new NotFoundEntityException("Esse plano não existe"));

        String userEmail = company.getLogin().getEmail();

        try {
            
            PreferenceItemRequest item = PreferenceItemRequest.builder()
                .title(currentPlan.getName())
                .description(currentPlan.getDescription())
                .quantity(1)
                .currencyId("BRL")
                .unitPrice(new BigDecimal(currentPlan.getPrice()))
                .build();



            PreferenceBackUrlsRequest backUrls = PreferenceBackUrlsRequest.builder()
                .success("https://localhost:5173")
                .failure("https://localhost:5173")
                .pending("https://localhost:5173")
                .build();

            List<PreferenceItemRequest> items = new ArrayList<>();
            items.add(item);

            PreferencePayerRequest payer = PreferencePayerRequest.builder()
                .name(company.getName())
                .email(userEmail)
                .build();

            
            Payment savedPayment = this.savePayment(company, currentPlan);

            PreferenceRequest preferenceRequest = PreferenceRequest.builder()
                .metadata(createMetadata(company.getExternalId(), userEmail, dto.planId(), savedPayment.getId()))
                .externalReference(company.getExternalId())
                .payer(payer)
                .backUrls(backUrls)
                .items(items)
                .autoReturn("approved")
                .build();

    
            PreferenceClient client = new PreferenceClient();
            Preference preference = client.create(preferenceRequest);

            if(preference.getId() == null){
                throw new MercadoPagoException("Não foi possível criar o pagamento");
            }

            MercadoPagoDTO mercadoPagoDTO = new MercadoPagoDTO(preference.getId(), preference.getInitPoint());

            return mercadoPagoDTO;

        } catch (Exception e) {
            System.out.println(e);
            System.out.println(e.getMessage());
            throw new MercadoPagoException("Não foi possível criar o pagamento");
        }
    }

    public void updatedPayment(Map<String, Object> body){


        String type = (String) body.get("type");

        try {
            if(type.equals("payment")){

                //data
                Map<String, Object> data = (Map<String, Object>) body.get("data");

                String paymentId = (String) data.get("id");

                PaymentClient client = new PaymentClient();
                com.mercadopago.resources.payment.Payment payment = client.get(Long.parseLong(paymentId));

                String collectorId = payment.getCollectorId().toString();
                String paymentTypeId = payment.getPaymentTypeId();
                String status = "approved";
                String externalReference = payment.getExternalReference();
                String paymentType = payment.getPaymentMethodId();
                String processingMode = payment.getProcessingMode();
                String merchantAccountId = payment.getMerchantAccountId();
                String confirmedDatePayment = payment.getDateApproved().toString();


                //Metadata
                Map<String, Object> metadata = payment.getMetadata();
                System.out.println(metadata);

                Double paymentIdDouble = (Double) metadata.get("payment_id");
                Long currentPaymentId = paymentIdDouble.longValue();
            
                Payment currentPayment = paymentRepository.findById(currentPaymentId).orElseThrow(() -> new NotFoundEntityException("Esse pagamento não existe"));
                currentPayment.setCollectorId(collectorId);;
                currentPayment.setPaymentId(paymentTypeId);
                currentPayment.setStatus(status);
                currentPayment.setExternalRefence(externalReference);
                currentPayment.setPaymentType(paymentType);
                currentPayment.setProcessingMode(processingMode);
                currentPayment.setMerchantAccountId(merchantAccountId);
                currentPayment.setConfirmedDatePayment(new Date());

                paymentRepository.save(currentPayment);
                String email = (String) metadata.get("user_email");
                this.updatedLogin(email);


                this.emailUtil.sendPaymentConfirmation(email);
                

            }
        } catch (Exception e) {
            System.out.println(e);
            System.out.println(e.getMessage());
            System.out.println("Caindo em error");
            throw new MercadoPagoException("Não foi possível atualizar o pagamento");
        }
    }

    private Map<String, Object> createMetadata(String companyId, String userEmail, Long planId, Long paymentId) {
        Map<String, Object> metadata = new HashMap<>();

        metadata.put("id", companyId);
        metadata.put("userEmail", userEmail);
        metadata.put("plan", planId);
        metadata.put("paymentId", paymentId);

        return metadata;
    }

    private Payment savePayment(Company company, Plan plan){

        Payment payment = new Payment("unpaid", plan, company);

        return paymentRepository.save(payment);
    }

    private void updatedLogin(String email){
        Login login = loginRepository.findByCompanyEmail(email);
        login.setActive(true);
        loginRepository.save(login);
    }
}
